rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -562,18 +562,35 @@ void Player::prepare_publishers()
   reader_->set_filter(storage_filter);
 
   // Create /clock publisher
-  if (play_options_.clock_publish_frequency > 0.f) {
-    const auto publish_period = std::chrono::nanoseconds(
-      static_cast<uint64_t>(RCUTILS_S_TO_NS(1) / play_options_.clock_publish_frequency));
+  if (play_options_.clock_publish_frequency > 0.f || play_options_.clock_publish_on_topic_publish) {
     // NOTE: PlayerClock does not own this publisher because rosbag2_cpp
     // should not own transport-based functionality
     clock_publisher_ = create_publisher<rosgraph_msgs::msg::Clock>(
       "/clock", rclcpp::ClockQoS());
+  }
+
+  if (play_options_.clock_publish_frequency > 0.f) {
+    const auto publish_period = std::chrono::nanoseconds(
+      static_cast<uint64_t>(RCUTILS_S_TO_NS(1) / play_options_.clock_publish_frequency));
+
     clock_publish_timer_ = create_wall_timer(
       publish_period, [this]() {
-        auto msg = rosgraph_msgs::msg::Clock();
-        msg.clock = rclcpp::Time(clock_->now());
-        clock_publisher_->publish(msg);
+        publish_clock_update();
+      });
+  }
+
+  if (play_options_.clock_publish_on_topic_publish) {
+    add_on_play_message_pre_callback(
+      [this](const std::shared_ptr<rosbag2_storage::SerializedBagMessage> msg) {
+        if (play_options_.clock_trigger_topics.empty()) {
+          publish_clock_update();
+        } else {
+          for (const auto & topic : play_options_.clock_trigger_topics) {
+            if (topic == msg->topic_name) {
+              publish_clock_update();
```

Comment:
@kylemarcey It doesn't make sense to iterate further if we have already found matching topic in list.
Can you please add `"break;"` from loop after `publish_clock_update();` ?  
This way we also will guarantee to publish only once even if topic name listed twice in the  `clock_trigger_topics`